# CMI - Detect Behavior With Sensor Data -- Initial Exploration

## Key Points From the Overview

### **Competition Details**

**The goal of this competition is to develop a predictive model that distinguishes BFRB-like and non-BFRB-like activity using data from a variety of sensors collected via a wrist-worn device. Successfully disentangling these behaviors will improve the design and accuracy of wearable BFRB-detection devices, which are relevant to a wide range of mental illnesses, ultimately strengthening the tools available to support their treatment.**

- To investigate BFRBs, the Child Mind Institute has developed a wrist-worn device, Helios, designed to detect these behaviors. While many commercially available devices contain Inertial Measurement Units (IMUs) to measure rotation and motion, the Helios watch integrates additional sensors, including 5 thermopiles (for detecting body heat) and 5 time-of-flight sensors (for detecting proximity). See the figure to the right for the placement of these sensors on the Helios device.

#### `Experiment`
- They began a transition from “rest” position and moved their hand to the appropriate location (Transition);
- They followed this with a short pause wherein they did nothing (Pause); and
- Finally they performed a gesture from either the BFRB-like or non-BFRB-like category of movements.
- Each participant performed 18 unique gestures (8 BFRB-like gestures and 10 non-BFRB-like gestures) in at least 1 of 4 different body-positions (sitting, sitting leaning forward with their non-dominant arm resting on their leg, lying on their back, and lying on their side).

#### `Challenge`

- This competition challenges you to develop a predictive model capable of distinguishing (1) BFRB-like gestures from non-BFRB-like gestures and (2) the specific type of BFRB-like gesture. 
- **Critically, when your model is evaluated, half of the test set will include only data from the IMU, while the other half will include all of the sensors on the Helios device (IMU, thermopiles, and time-of-flight sensors).**
- Your solutions will have direct real-world impact, as the insights gained will inform design decisions about sensor selection — specifically whether the added expense and complexity of thermopile and time-of-flight sensors is justified by significant improvements in BFRB detection accuracy compared to an IMU alone. 
    - By helping us determine the added value of these thermopiles and time-of-flight sensors, your work will guide the development of better tools for detection and treatment of BFRBs.


### **Data**

**`Types of Gestures`** 

- 8 BFRB
- 10 non- BFRB gestures

**`Helios Sensors + Hidden Test Set Info`**

- 1x Inertial Measurement Unit (IMU; BNO080/BNO085): An integrated sensor that combines accelerometer, gyroscope, and magnetometer measurements with onboard processing to provide orientation and motion data.
- 5x Thermopile Sensor (MLX90632): A non-contact temperature sensor that measures infrared radiation.
- 5x Time-of-Flight Sensor (VL53L7CX): A sensor that measures distance by detecting how long it takes for emitted infrared light to bounce back from objects.
- 3,500 sequences in the hidden test set.
- **Half of the hidden-test sequences are recorded with IMU only; the thermopile (thm_) and time-of-flight (tof__v*) columns are still present but contain null values for those sequences.**

**`Features / Columns to be Mindfull Of`**

- **sequence_type** target or non target gesture - train only.
- **gesture** `target column` - description of sequence gesture. train only.
- **orientation** subject orientation during sequence - train only.


## Initial EDA

In [1]:
from numba import cuda
print(cuda.gpus) # Test for GPU

<Managed Device 0>


In [9]:
from pathlib import Path
import pickle, gzip, os, time, shutil, subprocess, zipfile, gc
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import cudf # Consistent dataframe merge errors revert to pandas for EDA and FE

warnings.filterwarnings('ignore')
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (12, 8)

In [3]:
def display_all(df):
    with pd.option_context("display.max_rows", 100, "display.max_columns", 1000):
        display(df)

In [4]:
## Move to utils
def download_data(competition:str, base_path: Path):
    data_path = base_path/'data'
    zip_path = data_path / f"{competition}.zip"
    
    data_path.mkdir(parents=True, exist_ok=False)
    print(f"Downloading data for competition: {competition}...")
    try:
        command = [
            "kaggle", "competitions", "download",
            "-c", competition,
            "-p", str(data_path),
            "--force"
            ]
        result = subprocess.run(
            command,
            check=True,
            text=True,
            encoding='utf-8'
        )
        print("Kaggle API Output: ", result.stdout)
    except FileNotFoundError: 
        print("'kaggle' command not found")
        return
    except subprocess.CalledProcessError as e:
        print(f"Error downloading data. Kaggle returned an API error: {e}")
        return

    try:
        with zipfile.ZipFile(zip_path, 'r') as z:
            z.extractall(data_path)
        print("Unzipping Complete.")
    except zipfile.BadZipFile:
        print(f"Error: Failed to unzip {zip_path}. The file may be corrupted.")
        return
    
    os.remove(zip_path)
        

In [5]:
competition_name = "cmi-detect-behavior-with-sensor-data"
project_dir = Path(os.path.expanduser("~/code/kaggle/kaggle-cmi-detect-behavior"))

#download_data(competition_name, project_dir)

Loading training data...

In [16]:
train_path = project_dir/'data/train.csv'
train_demo_path = project_dir/'data/train_demographics.csv'
test_path = project_dir/'data/test.csv'
test_demo_path =project_dir/'data/test_demographics.csv'

#Load DFs
train = pd.read_csv(train_path)
train_demo = pd.read_csv(train_demo_path)

### Analyzing `train`

In [17]:
display_all(train.head())

row_id sequence_type sequence_id  sequence_counter      subject  \
0  SEQ_000007_000000        Target  SEQ_000007                 0  SUBJ_059520   
1  SEQ_000007_000001        Target  SEQ_000007                 1  SUBJ_059520   
2  SEQ_000007_000002        Target  SEQ_000007                 2  SUBJ_059520   
3  SEQ_000007_000003        Target  SEQ_000007                 3  SUBJ_059520   
4  SEQ_000007_000004        Target  SEQ_000007                 4  SUBJ_059520   

                       orientation                                   behavior  \
0  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
1  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
2  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
3  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
4  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   

        phase             gesture     acc_x     acc_y     acc_z     rot_w  \
0  Transition  Cheek - pinch skin  6.683594  6.214844  3.355469  0.134399   
1  Transition  Cheek - pinch skin  6.949219  6.214844  3.125000  0.143494   
2  Transition  Cheek - pinch skin  5.722656  5.410156  5.421875  0.219055   
3  Transition  Cheek - pinch skin  6.601562  3.531250  6.457031  0.297546   
4  Transition  Cheek - pinch skin  5.566406  0.277344  9.632812  0.333557   

      rot_x     rot_y     rot_z      thm_1      thm_2      thm_3      thm_4  \
0 -0.355164 -0.447327 -0.809753  28.943842  31.822186  29.553024  28.592863   
1 -0.340271 -0.428650 -0.824524  29.340816  31.874645  29.791740  28.663383   
2 -0.274231 -0.356934 -0.865662  30.339359  30.935045  30.090014  28.796087   
3 -0.264160 -0.238159 -0.885986  30.543730  27.044001  29.310717  29.018711   
4 -0.218628 -0.063538 -0.914856  29.317265  25.270855  26.808746  29.408604   

       thm_5  tof_1_v0  tof_1_v1  tof_1_v2  tof_1_v3  tof_1_v4  tof_1_v5  \
0  28.310535     131.0     134.0     132.0     135.0      98.0      74.0   
1  28.406172     130.0     138.0     131.0     135.0     101.0      76.0   
2  28.529778     137.0     136.0     147.0     109.0      90.0      81.0   
3  27.402010     143.0     147.0     170.0     127.0     109.0      98.0   
4  27.357603     178.0     191.0     183.0     157.0     146.0     139.0   

   tof_1_v6  tof_1_v7  tof_1_v8  tof_1_v9  tof_1_v10  tof_1_v11  tof_1_v12  \
0      64.0      60.0      -1.0      -1.0      152.0      153.0      141.0   
1      66.0      61.0      -1.0      -1.0      156.0      155.0      141.0   
2      74.0      74.0      -1.0     164.0      165.0      146.0      106.0   
3      95.0      95.0      -1.0     177.0      189.0      177.0      136.0   
4     143.0     148.0      -1.0      -1.0      236.0      238.0      208.0   

   tof_1_v13  tof_1_v14  tof_1_v15  tof_1_v16  tof_1_v17  tof_1_v18  \
0       89.0       68.0       63.0       -1.0       -1.0       -1.0   
1       93.0       74.0       64.0       -1.0       -1.0       -1.0   
2       94.0       77.0       77.0       -1.0       -1.0       -1.0   
3      121.0      107.0      104.0       -1.0       -1.0       -1.0   
4      200.0      185.0      190.0       -1.0       -1.0       -1.0   

   tof_1_v19  tof_1_v20  tof_1_v21  tof_1_v22  tof_1_v23  tof_1_v24  \
0       -1.0      169.0      118.0       86.0       73.0       -1.0   
1       -1.0      165.0      116.0       86.0       75.0      130.0   
2      180.0      140.0      118.0      103.0       92.0       -1.0   
3      202.0      171.0      160.0      141.0      135.0       -1.0   
4      210.0      246.0      225.0      228.0      202.0      149.0   

   tof_1_v25  tof_1_v26  tof_1_v27  tof_1_v28  tof_1_v29  tof_1_v30  \
0       -1.0       -1.0       -1.0       -1.0      147.0      110.0   
1       -1.0       -1.0       -1.0       -1.0      142.0      114.0   
2       -1.0       -1.0       -1.0       -1.0      155.0      119.0   
3       -1.0       -1.0       -1.0    

In [18]:
# Get null percentages per column
null_info = [(col, train[col].isnull().sum(), (train[col].isnull().sum() / len(train))*100)
             for col in train.columns if train[col].isnull().any()]

for col, cnt, perc in null_info:
    print(f"{col}: {cnt} nulls ({perc:.2f}%)")
print(train.shape)

rot_w: 3692 nulls (0.64%)
rot_x: 3692 nulls (0.64%)
rot_y: 3692 nulls (0.64%)
rot_z: 3692 nulls (0.64%)
thm_1: 6987 nulls (1.22%)
thm_2: 7638 nulls (1.33%)
thm_3: 6472 nulls (1.13%)
thm_4: 6224 nulls (1.08%)
thm_5: 33286 nulls (5.79%)
tof_1_v0: 6224 nulls (1.08%)
tof_1_v1: 6224 nulls (1.08%)
tof_1_v2: 6224 nulls (1.08%)
tof_1_v3: 6224 nulls (1.08%)
tof_1_v4: 6224 nulls (1.08%)
tof_1_v5: 6224 nulls (1.08%)
tof_1_v6: 6224 nulls (1.08%)
tof_1_v7: 6224 nulls (1.08%)
tof_1_v8: 6224 nulls (1.08%)
tof_1_v9: 6224 nulls (1.08%)
tof_1_v10: 6224 nulls (1.08%)
tof_1_v11: 6224 nulls (1.08%)
tof_1_v12: 6224 nulls (1.08%)
tof_1_v13: 6224 nulls (1.08%)
tof_1_v14: 6224 nulls (1.08%)
tof_1_v15: 6224 nulls (1.08%)
tof_1_v16: 6224 nulls (1.08%)
tof_1_v17: 6224 nulls (1.08%)
tof_1_v18: 6224 nulls (1.08%)
tof_1_v19: 6224 nulls (1.08%)
tof_1_v20: 6224 nulls (1.08%)
tof_1_v21: 6224 nulls (1.08%)
tof_1_v22: 6224 nulls (1.08%)
tof_1_v23: 6224 nulls (1.08%)
tof_1_v24: 6224 nulls (1.08%)
tof_1_v25: 6224 nulls (1.

In [19]:
# Unique counts per sequence
print(f"Number of unique sequences: {train['sequence_id'].nunique()}")
print(f"Number of unique subjects : {train['subject'].nunique()}")
print(f"Number of unique gestures(target): {train['gesture'].nunique()}")

Number of unique sequences: 8151
Number of unique subjects : 81
Number of unique gestures(target): 18


In [29]:
display_all(train.describe())

sequence_counter          acc_x          acc_y          acc_z  \
count     574945.000000  574945.000000  574945.000000  574945.000000   
mean          43.645234       1.639980       1.790704      -0.459811   
std           45.717673       5.781259       5.003945       6.096490   
min            0.000000     -34.585938     -24.402344     -42.855469   
25%           17.000000      -2.964844      -2.121094      -5.417969   
50%           35.000000       2.972656       0.695312      -1.562500   
75%           54.000000       6.343750       6.816406       5.164062   
max          699.000000      46.328125      27.183594      30.078125   

               rot_w          rot_x          rot_y          rot_z  \
count  571253.000000  571253.000000  571253.000000  571253.000000   
mean        0.360375      -0.119916      -0.059953      -0.188298   
std         0.225739       0.465520       0.543028       0.504137   
min         0.000000      -0.999146      -0.999695      -0.998169   
25%         0.180237      -0.456299      -0.511536      -0.627686   
50%         0.340332      -0.186890      -0.112610      -0.263916   
75%         0.503479       0.204590       0.440063       0.251099   
max         0.999390       0.999817       0.999451       0.999878   

               thm_1          thm_2          thm_3          thm_4  \
count  567958.000000  567307.000000  568473.000000  568721.000000   
mean       27.076450      27.133483      26.702994      27.559114   
std         3.231948       2.941438       4.122353       2.253639   
min        -0.370413      21.958820       0.000000      22.375921   
25%        24.753527      24.543737      24.640350      25.594980   
50%        26.982324      26.354338      26.956276      27.742224   
75%        29.425036      29.620147      29.231794      29.267954   
max        38.457664      37.578339      37.294994      39.588390   

               thm_5       tof_1_v0       tof_1_v1       tof_1_v2  \
count  541659.000000  568721.000000  568721.000000  568721.000000   
mean       26.669643      54.591513      54.997108      56.233888   
std         2.438397      72.756604      73.209203      74.045765   
min        22.045000      -1.000000      -1.000000      -1.000000   
25%        24.556606      -1.000000      -1.000000      -1.000000   
50%        26.185198      -1.000000      -1.000000      -1.000000   
75%        28.574785     104.000000     106.000000     110.000000   
max        37.683292     249.000000     249.000000     249.000000   

            tof_1_v3       tof_1_v4       tof_1_v5       tof_1_v6  \
count  568721.000000  568721.000000  568721.000000  568721.000000   
mean       57.663855      58.766488      57.383394      56.358724   
std        74.668750      74.512695      72.857238      71.116937   
min        -1.000000      -1.000000      -1.000000      -1.000000   
25%        -1.000000      -1.000000      -1.000000      -1.000000   
50%        -1.000000       0.000000       3.000000       9.000000   
75%       113.000000     115.000000     111.000000     108.000000   
max       249.000000     249.000000     249.000000     249.000000   

            tof_1_v7       tof_1_v8       tof_1_v9      tof_1_v10  \
count  568721.000000  568721.000000  568721.000000  568721.000000   
mean       56.353711      49.743419      52.699923      52.970339   
std        70.139115      71.121335      73.522558      73.790710   
min        -1.000000      -1.000000      -1.000000      -1.000000   
25%        -1.000000      -1.000000      -1.000000      -1.000000   
50%        19.000000      -1.000000      -1.000000      -1.000000   
75%       106.000000      94.000000     103.000000     104.000000   
max       249.000000     249.000000     249.000000     249.000000   

           tof_1_v11      tof_1_v12      tof_1_v13      tof_1_v14  \
count  568721.000000  568721.000000  568721.000000  568721.000000   
mean       54.872648      56.163426      55.645849      53.562255   
std        74.534383      74.833642      

#### Initial takeaways

- With 81 individual subjects and 18 unique gestures per subject, this is a non-trivial multi-class problem.
- Looking at NaN values per column, there appear to be two distinct types of sensor failures:
    - **Intermittent Failures** `rot_*` [quarternions](https://en.wikipedia.org/wiki/Quaternion), `thm_1-4` and `tof_1-4` sensors exhibit low and consistent NaN percentages (~0.64 and ~1.08%). Could this possibly be because of sporadic, temporary loss of communication from specific sensor chips?
    - **Systematic Failure** `thm_5` and all of `tof_5` pixels have a much higher and identical failure rate (~5.5%). Suggesting, quite strongly, that the fifth group of sensors is less reliable or could have been partially disabled on some devices.
- For many of the `tof_*` pixels the 25th, and sometimes the 50th, percentiles are -1.
    - This indicates that -1 is not an outlier but a frequent state representing "no object detected in range". This is fundamental to the data.
    - The value must be treated carefully. During feature engineering statistics need to be calculated for both with -1 and without -1 values.
- Moving to categorical variables, the `phase` column (Transition, Pause and Gesture) segments the behaviour within a sequence.

### Analyzing `test demographics`

In [21]:
train_demo.head()

subject  adult_child  age  sex  handedness  height_cm  \
0  SUBJ_000206            1   41    1           1      172.0   
1  SUBJ_001430            0   11    0           1      167.0   
2  SUBJ_002923            1   28    1           0      164.0   
3  SUBJ_003328            1   33    1           1      171.0   
4  SUBJ_004117            0   15    0           1      184.0   

   shoulder_to_wrist_cm  elbow_to_wrist_cm  
0                    50               25.0  
1                    51               27.0  
2                    54               26.0  
3                    52               25.0  
4                    54               28.0

In [22]:
train_demo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 81 entries, 0 to 80
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   subject               81 non-null     object 
 1   adult_child           81 non-null     int64  
 2   age                   81 non-null     int64  
 3   sex                   81 non-null     int64  
 4   handedness            81 non-null     int64  
 5   height_cm             81 non-null     float64
 6   shoulder_to_wrist_cm  81 non-null     int64  
 7   elbow_to_wrist_cm     81 non-null     float64
dtypes: float64(2), int64(5), object(1)
memory usage: 5.2+ KB


In [23]:
train_demo.describe()

adult_child        age        sex  handedness   height_cm  \
count    81.000000  81.000000  81.000000   81.000000   81.000000   
mean      0.518519  21.814815   0.617284    0.876543  167.993827   
std       0.502770  10.289693   0.489078    0.331010   10.605863   
min       0.000000  10.000000   0.000000    0.000000  135.000000   
25%       0.000000  13.000000   0.000000    1.000000  163.000000   
50%       1.000000  22.000000   1.000000    1.000000  170.000000   
75%       1.000000  27.000000   1.000000    1.000000  174.000000   
max       1.000000  53.000000   1.000000    1.000000  190.500000   

       shoulder_to_wrist_cm  elbow_to_wrist_cm  
count             81.000000          81.000000  
mean              51.580247          25.469136  
std                4.888415           3.025254  
min               41.000000          18.000000  
25%               49.000000          24.000000  
50%               52.000000          25.000000  
75%               55.000000          27.000000  
max               71.000000          44.000000

In [24]:
# Basic distributions
print(f"Sex distribution : {train_demo['sex'].value_counts()}") # female=0 -- male=1
print(f"Child / Adult distribution : {train_demo['adult_child'].value_counts()}") # child=0 -- adult=1
print(f"Handedness distribution : {train_demo['handedness'].value_counts()}") # left handed=0 -- right handed=1

Sex distribution : sex
1    50
0    31
Name: count, dtype: int64
Child / Adult distribution : adult_child
1    42
0    39
Name: count, dtype: int64
Handedness distribution : handedness
1    71
0    10
Name: count, dtype: int64


#### Initial takeaways

- There is a reasonable split between adults and children.
- The data is skewed towards males and right handed participants.

## Merge Training Data Sources For Deeper EDA

In [14]:
# Merging sensor training data with demographics data
train_full = train.merge(train_demo, left_on='subject', right_on='subject', how='left')

print(f"Shape of merged training data: {train_full.shape}")
print("Columns: ", train_full.columns.to_list())

Shape of merged training data: (574945, 348)
Columns:  ['row_id', 'sequence_type', 'sequence_id', 'sequence_counter', 'subject', 'orientation', 'behavior', 'phase', 'gesture', 'acc_x', 'acc_y', 'acc_z', 'rot_w', 'rot_x', 'rot_y', 'rot_z', 'thm_1', 'thm_2', 'thm_3', 'thm_4', 'thm_5', 'tof_1_v0', 'tof_1_v1', 'tof_1_v2', 'tof_1_v3', 'tof_1_v4', 'tof_1_v5', 'tof_1_v6', 'tof_1_v7', 'tof_1_v8', 'tof_1_v9', 'tof_1_v10', 'tof_1_v11', 'tof_1_v12', 'tof_1_v13', 'tof_1_v14', 'tof_1_v15', 'tof_1_v16', 'tof_1_v17', 'tof_1_v18', 'tof_1_v19', 'tof_1_v20', 'tof_1_v21', 'tof_1_v22', 'tof_1_v23', 'tof_1_v24', 'tof_1_v25', 'tof_1_v26', 'tof_1_v27', 'tof_1_v28', 'tof_1_v29', 'tof_1_v30', 'tof_1_v31', 'tof_1_v32', 'tof_1_v33', 'tof_1_v34', 'tof_1_v35', 'tof_1_v36', 'tof_1_v37', 'tof_1_v38', 'tof_1_v39', 'tof_1_v40', 'tof_1_v41', 'tof_1_v42', 'tof_1_v43', 'tof_1_v44', 'tof_1_v45', 'tof_1_v46', 'tof_1_v47', 'tof_1_v48', 'tof_1_v49', 'tof_1_v50', 'tof_1_v51', 'tof_1_v52', 'tof_1_v53', 'tof_1_v54', 'tof_1_v55'

In [25]:
display_all(train_full.head())

row_id sequence_type sequence_id  sequence_counter      subject  \
0  SEQ_000007_000000        Target  SEQ_000007                 0  SUBJ_059520   
1  SEQ_000007_000001        Target  SEQ_000007                 1  SUBJ_059520   
2  SEQ_000007_000002        Target  SEQ_000007                 2  SUBJ_059520   
3  SEQ_000007_000003        Target  SEQ_000007                 3  SUBJ_059520   
4  SEQ_000007_000004        Target  SEQ_000007                 4  SUBJ_059520   

                       orientation                                   behavior  \
0  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
1  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
2  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
3  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
4  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   

        phase             gesture     acc_x     acc_y     acc_z     rot_w  \
0  Transition  Cheek - pinch skin  6.683594  6.214844  3.355469  0.134399   
1  Transition  Cheek - pinch skin  6.949219  6.214844  3.125000  0.143494   
2  Transition  Cheek - pinch skin  5.722656  5.410156  5.421875  0.219055   
3  Transition  Cheek - pinch skin  6.601562  3.531250  6.457031  0.297546   
4  Transition  Cheek - pinch skin  5.566406  0.277344  9.632812  0.333557   

      rot_x     rot_y     rot_z      thm_1      thm_2      thm_3      thm_4  \
0 -0.355164 -0.447327 -0.809753  28.943842  31.822186  29.553024  28.592863   
1 -0.340271 -0.428650 -0.824524  29.340816  31.874645  29.791740  28.663383   
2 -0.274231 -0.356934 -0.865662  30.339359  30.935045  30.090014  28.796087   
3 -0.264160 -0.238159 -0.885986  30.543730  27.044001  29.310717  29.018711   
4 -0.218628 -0.063538 -0.914856  29.317265  25.270855  26.808746  29.408604   

       thm_5  tof_1_v0  tof_1_v1  tof_1_v2  tof_1_v3  tof_1_v4  tof_1_v5  \
0  28.310535     131.0     134.0     132.0     135.0      98.0      74.0   
1  28.406172     130.0     138.0     131.0     135.0     101.0      76.0   
2  28.529778     137.0     136.0     147.0     109.0      90.0      81.0   
3  27.402010     143.0     147.0     170.0     127.0     109.0      98.0   
4  27.357603     178.0     191.0     183.0     157.0     146.0     139.0   

   tof_1_v6  tof_1_v7  tof_1_v8  tof_1_v9  tof_1_v10  tof_1_v11  tof_1_v12  \
0      64.0      60.0      -1.0      -1.0      152.0      153.0      141.0   
1      66.0      61.0      -1.0      -1.0      156.0      155.0      141.0   
2      74.0      74.0      -1.0     164.0      165.0      146.0      106.0   
3      95.0      95.0      -1.0     177.0      189.0      177.0      136.0   
4     143.0     148.0      -1.0      -1.0      236.0      238.0      208.0   

   tof_1_v13  tof_1_v14  tof_1_v15  tof_1_v16  tof_1_v17  tof_1_v18  \
0       89.0       68.0       63.0       -1.0       -1.0       -1.0   
1       93.0       74.0       64.0       -1.0       -1.0       -1.0   
2       94.0       77.0       77.0       -1.0       -1.0       -1.0   
3      121.0      107.0      104.0       -1.0       -1.0       -1.0   
4      200.0      185.0      190.0       -1.0       -1.0       -1.0   

   tof_1_v19  tof_1_v20  tof_1_v21  tof_1_v22  tof_1_v23  tof_1_v24  \
0       -1.0      169.0      118.0       86.0       73.0       -1.0   
1       -1.0      165.0      116.0       86.0       75.0      130.0   
2      180.0      140.0      118.0      103.0       92.0       -1.0   
3      202.0      171.0      160.0      141.0      135.0       -1.0   
4      210.0      246.0      225.0      228.0      202.0      149.0   

   tof_1_v25  tof_1_v26  tof_1_v27  tof_1_v28  tof_1_v29  tof_1_v30  \
0       -1.0       -1.0       -1.0       -1.0      147.0      110.0   
1       -1.0       -1.0       -1.0       -1.0      142.0      114.0   
2       -1.0       -1.0       -1.0       -1.0      155.0      119.0   
3       -1.0       -1.0       -1.0    